### DISCUSSION TITLE
Using C-Blocks for object scope?

### DISCUSSION THREAD

**Discussion by reesericci** [TurboWarp/extensions#1403] [General]
Coming from Ruby, blocks are used plenty for controlling scope, and I think it makes it clear what objects are in use. Here's an example Rails database migration:
```
create_table :products do |t|
      t.string :name
      t.text :description

      t.timestamps
    end
 ``` 
 
 This puts the t object, the products table, in scope for that block, which I think is cleaner instead of a series of commands like `add_column :products :name :string` where the column needs to be specified every time, or something like `switch_table` (fictional) before running commands.
 
 Coming back to Scratch, I find myself wanting to do similar scope controls while working on turbowarp.sh. Here's an example: I want to have it be possible to use multiple ttys "contexts" for running commands within one script. I briefly considered starting a new context on every script, like an event block, but that doesn't make sense. I could add a switch tty block, but it's just not as elegant as what a c block would be. Here's a picture with comment blocks (it also has some new expect stuff i'm working on):
 
 
![image](https://github.com/TurboWarp/extensions/assets/19589006/44691706-c5f3-481b-84bd-4262526993f3)

This allows for a few things:

1) Allowing for extraneous scripts to run independently of a global context, like the apt install - but a better example is if it was flipped, where the ls was in the C-block and the apt install was running globally.

2) Discarding the TTY at the end of the script automatically, instead of doing it with a when stop button click or just leaving it open until a new connection

3) Making it incredibly clear which TTY is in use, and reducing shared state

What are y'alls thoughts on this?

Thanks,

--reese

**Comment by AshimeeAlt** [comment]
if you want stuff to run independently from the main script use pushThread

**Reply by reesericci** [reply]
That's not the goal, this is talking about closure & scope https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures

**Reply by AshimeeAlt** [reply]
you cant do this, and this would not be good for scratch / tw itself

**Reply by reesericci** [reply]
I think reducing shared state with scope would be good

**Reply by AshimeeAlt** [reply]
still, if you need "scope" in tw you should not be using it and if you just need it for variables then scope variable extensions exist
